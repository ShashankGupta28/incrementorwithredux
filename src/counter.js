import React from 'react';
import { connect } from 'react-redux';
import * as CONSTANT_VALUES from './constant';
import * as ACTION_TYPE from './action/actionType';

export class Counter extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      count: CONSTANT_VALUES.MIN_COUNTER_VALUE
    }
    this.increment = this.increment.bind(this);
    this.onChange = this.onChange.bind(this);
  }
  //For receiving updated store values
  componentWillReceiveProps(nextProps) {
    var state = this.state;
    state.count = (nextProps && nextProps.count) ? nextProps.count : state.count;
    this.setState(state);
  }
  // For changing input of counter & validating counter value should not be greater than 100
  onChange(e) {
    var state = this.state;
    if (e.target.value === '' || (CONSTANT_VALUES.COUNTER_REGEX).test(e.target.value)) {
      state.count = (Number(e.target.value) > CONSTANT_VALUES.MAX_COUNTER_VALUE) ? CONSTANT_VALUES.MAX_COUNTER_VALUE : Number(e.target.value);
      this.setState(state);
      this.props.dispatch({ type: ACTION_TYPE.ONCHANGE, counter: state.count });

    }

  }
  // For incrementing counter value by 1
  increment() {
    if (this.state.count < CONSTANT_VALUES.MAX_COUNTER_VALUE) {
      this.props.dispatch({ type: ACTION_TYPE.INCREMENT, counter: this.state.count });
    }
  }
  // For decrementing counter value by 1
  decrement = () => {
    var state = this.state;
    if (this.state.count > CONSTANT_VALUES.MIN_COUNTER_VALUE && this.state.count !== CONSTANT_VALUES.MIN2_COUNTER_VALUE) {
      this.props.dispatch({ type: ACTION_TYPE.DECREMENT, counter: this.state.count });
    }
    else if (this.state.count === CONSTANT_VALUES.MIN2_COUNTER_VALUE) {
      state.count = CONSTANT_VALUES.MIN_COUNTER_VALUE;
      this.setState(state);
      this.props.dispatch({ type: ACTION_TYPE.ONCHANGE, counter: CONSTANT_VALUES.MIN_COUNTER_VALUE });
    }
  }
  render() {
    return (
      <div id="outer-div">
        <h2>Counter</h2>
        <div>
          <button id="decrement" onClick={this.decrement}>-</button>
          <input type="text" id="counter" value={this.state.count} onChange={this.onChange} />
          <button id="increment" onClick={this.increment}>+</button>
        </div>
      </div>
    )
  }
}

function mapStateToProps(state) {
  console.log("map state to", state.counterReducer.count, state);
  return {
    count: state.counterReducer.count
  };
}

export default connect(mapStateToProps)(Counter);