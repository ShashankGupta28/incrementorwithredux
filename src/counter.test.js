import React from "react";
import {Counter} from './counter';
import renderer from "react-test-renderer";
import { mount } from "enzyme";
// import { configure,mount } from 'enzyme';
// import Adapter from 'enzyme-adapter-react-15';

// configure({ adapter: new Adapter() });;


describe("Counter", () => {
    it("renders xcorrectly", () => {
        const output = renderer
      .create(<div id="outer-div">
      <h2>Counter</h2>
      <div>
        <button >-</button>
        <input type="text"  />
        <button >+</button>
      </div>
    </div>)
      .toJSON();
    expect(output).toMatchSnapshot();
    });

  //   it("calls on change on click", () => {
  //     const onChangeSpy = jest.fn();
      
  //     const wrapper = mount(<Counter/>);
  //   wrapper.find("input").first().simulate("change");
  //   expect(onChangeSpy).toHaveBeenCalled();
  // });



  

});