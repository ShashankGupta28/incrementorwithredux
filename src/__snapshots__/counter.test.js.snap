// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter renders correctly 1`] = `
<div
  id="outer-div"
>
  <h2>
    Counter
  </h2>
  <div>
    <button>
      -
    </button>
    <input
      type="text"
    />
    <button>
      +
    </button>
  </div>
</div>
`;

exports[`Counter renders xcorrectly 1`] = `
<div
  id="outer-div"
>
  <h2>
    Counter
  </h2>
  <div>
    <button>
      -
    </button>
    <input
      type="text"
    />
    <button>
      +
    </button>
  </div>
</div>
`;
